File: lib/enrich/driver.ts
// Enrichment driver: walks a dump, runs the per-type enrichers with a bounded
// concurrency pool, and writes _enrich_report.json. Logic moved from
// enrich_bodies.ts (the resumability gate, stale-key clearing, per-type counters,
// progress logging) with two injection changes: an HttpClient is passed in (so
// the GitHub token is read in the cmd layer, not here) and a Logger replaces the
// bare console.* calls. Behavior is otherwise unchanged.

import { type Logger, NULL_LOGGER } from "../logger.ts";
import { HttpClient } from "../http/fetcher.ts";
import { exists, readJson, walkArticleFiles, writeJson } from "../fsutil.ts";
import type { Changelog } from "../changelog.ts";
import { mergePending, pendingDir, type PendingEntry, pendingPath } from "../staging.ts";
import {
  type Article,
  type EnricherDeps,
  type EnrichResult,
  hasBody,
  TYPE_ENRICHERS,
} from "./enrichers.ts";

export interface TypeReport {
  typeKey: string;
  files: number;
  enriched: number;
  failed: number;
  skipped: number;
  /** approval gate: re-fetches that would overwrite a live body were staged instead. */
  staged: number;
  missingDir?: boolean;
  errors: Array<{ id: string; link: string; error: string }>;
}

const defaultSleep = (ms: number) => new Promise<void>((r) => setTimeout(r, ms));

// Run async tasks with a fixed concurrency, returning when all are done.
export async function runPool<T>(
  items: T[],
  concurrency: number,
  worker: (item: T) => Promise<void>,
): Promise<void> {
  let idx = 0;
  const runners = Array.from({ length: Math.min(concurrency, items.length) }, async () => {
    while (idx < items.length) {
      const i = idx++;
      await worker(items[i]);
    }
  });
  await Promise.all(runners);
}

export interface EnrichTypeOpts {
  typeKey: string;
  dump: string;
  http: HttpClient;
  githubToken?: string;
  concurrency: number;
  delayMs: number;
  limit: number | null;
  refetch: boolean;
  refetchErrors: boolean;
  logger?: Logger;
  sleep?: (ms: number) => Promise<void>;
  changelog?: Changelog;
  /** approval gate: a --refetch that would overwrite an existing body is written to
   *  _pending/ instead of the live file. */
  approval?: boolean;
  /** collector for staged entries (the caller merges them into the manifest). */
  pending?: PendingEntry[];
}

async function listArticleFiles(dir: string): Promise<string[]> {
  const files: string[] = [];
  for await (const file of walkArticleFiles(dir)) files.push(file);
  files.sort();
  return files;
}

export async function enrichType(opts: EnrichTypeOpts): Promise<TypeReport> {
  const logger = opts.logger ?? NULL_LOGGER;
  const sleep = opts.sleep ?? defaultSleep;
  const { typeKey } = opts;
  const report: TypeReport = {
    typeKey,
    files: 0,
    enriched: 0,
    failed: 0,
    skipped: 0,
    staged: 0,
    errors: [],
  };
  const enricher = TYPE_ENRICHERS[typeKey];
  const deps: EnricherDeps = { http: opts.http, githubToken: opts.githubToken };
  const dir = `${opts.dump}/${typeKey}`;
  let files: string[];
  try {
    files = await listArticleFiles(dir);
  } catch {
    logger.info(`  [${typeKey}] no directory ${dir} — skipping`);
    report.missingDir = true;
    return report;
  }
  if (opts.limit) files = files.slice(0, opts.limit);
  report.files = files.length;

  let done = 0, skipped = 0, ok = 0, failed = 0, staged = 0;
  const nowIso = new Date().toISOString();

  await runPool(files, opts.concurrency, async (file) => {
    const article = await readJson<Article>(file);
    const hadError = typeof article.content?.bodyError === "string";
    // Skip already-done articles unless forced. --refetch-errors re-processes
    // only those that previously errored (e.g. after mapping a new host).
    if (!opts.refetch && !(opts.refetchErrors && hadError) && hasBody(article.content)) {
      skipped++;
      return;
    }
    const hadBodyBefore = hasBody(article.content);
    let result: EnrichResult;
    try {
      result = await enricher(article, nowIso, deps);
      ok++;
    } catch (e) {
      const error = (e as Error).message;
      result = { bodySource: article.link ?? "", fetchedAt: nowIso, bodyError: error };
      failed++;
      report.errors.push({ id: article.id ?? "", link: article.link ?? "", error });
    }
    opts.changelog?.record({
      op: result.bodyError ? "body-error" : (hadBodyBefore ? "body-changed" : "body-added"),
      documentType: article.documentType ?? opts.typeKey,
      id: article.id ?? "",
      title: article.title,
      source: "enrich",
      detail: result.bodyError,
    });
    // Clear any keys a prior run set so a re-enrich (e.g. after fixing a
    // JS-rendered host) doesn't leave a stale bodyError/body_text behind.
    const base = { ...(article.content ?? {}) };
    for (const k of ["sections", "body_text", "bodyError", "bodySource", "fetchedAt"]) {
      delete (base as Record<string, unknown>)[k];
    }
    article.content = { ...base, ...result };
    // Approval gate: a --refetch that would OVERWRITE an existing live body is
    // routed to _pending/ instead, leaving the good live file for review. Articles
    // with no prior body (new fills, error-stub retries) are written in place —
    // there is no good data to lose.
    if (opts.approval && opts.refetch && hadBodyBefore) {
      const pp = pendingPath(opts.dump, typeKey, article.id ?? "");
      await Deno.mkdir(pp.slice(0, pp.lastIndexOf("/")), { recursive: true });
      await writeJson(pp, article);
      opts.pending?.push({
        typeKey,
        id: article.id ?? "",
        title: article.title,
        op: "edited",
        source: "enrich",
        stagedAt: nowIso,
      });
      staged++;
    } else {
      await writeJson(file, article);
    }
    done++;
    if ((done + skipped) % 25 === 0) {
      logger.info(
        `  [${typeKey}] ${
          done + skipped
        }/${files.length} (ok=${ok} fail=${failed} skip=${skipped})`,
      );
    }
    if (opts.delayMs) await sleep(opts.delayMs);
  });

  report.enriched = ok;
  report.failed = failed;
  report.skipped = skipped;
  report.staged = staged;
  const stg = staged ? ` staged=${staged}` : "";
  logger.info(
    `  [${typeKey}] DONE: ${files.length} files — enriched=${ok} failed=${failed} skipped=${skipped}${stg}`,
  );
  return report;
}

export interface EnrichDumpOpts {
  dump: string;
  types?: string[] | null;
  http: HttpClient;
  /** Read in the cmd layer (from env), never here; passed through to the github enricher. */
  githubToken?: string;
  concurrency: number;
  delayMs: number;
  limit: number | null;
  refetch: boolean;
  refetchErrors: boolean;
  logger?: Logger;
  sleep?: (ms: number) => Promise<void>;
  changelog?: Changelog;
  /** approval gate: stage --refetch overwrites instead of clobbering live bodies. */
  approval?: boolean;
}

export async function enrichDump(opts: EnrichDumpOpts): Promise<TypeReport[]> {
  const logger = opts.logger ?? NULL_LOGGER;
  const requested = opts.types ?? Object.keys(TYPE_ENRICHERS);
  const toRun = requested.filter((t) => {
    if (!TYPE_ENRICHERS[t]) {
      logger.error(`  [${t}] no enricher implemented — skipping`);
      return false;
    }
    return true;
  });
  if (toRun.length === 0) {
    throw new Error("Nothing to do. Implemented types: " + Object.keys(TYPE_ENRICHERS).join(", "));
  }
  logger.info(`Enriching bodies in ${opts.dump} for: ${toRun.join(", ")}`);
  logger.info(
    `(concurrency=${opts.concurrency}, delay=${opts.delayMs}ms, refetch=${opts.refetch})`,
  );
  if (toRun.includes("F5_GitHub")) {
    logger.info(
      `GitHub auth: ${
        opts.githubToken
          ? "token present (5000/hr)"
          : "UNAUTHENTICATED (60/hr) — set GITHUB_TOKEN to raise"
      }`,
    );
  }
  const staged: PendingEntry[] = [];
  const reports: TypeReport[] = [];
  for (const t of toRun) {
    reports.push(
      await enrichType({
        typeKey: t,
        dump: opts.dump,
        http: opts.http,
        githubToken: opts.githubToken,
        concurrency: opts.concurrency,
        delayMs: opts.delayMs,
        limit: opts.limit,
        refetch: opts.refetch,
        refetchErrors: opts.refetchErrors,
        logger: opts.logger,
        sleep: opts.sleep,
        changelog: opts.changelog,
        approval: opts.approval,
        pending: staged,
      }),
    );
  }

  // Fill bodies of articles that an earlier dump/sync STAGED into _pending/ (so a
  // reviewer sees the complete new article, body included, before approving). This
  // pass is never gated — _pending/ is a staging area, not protected live data.
  if (await exists(pendingDir(opts.dump))) {
    for (const t of toRun) {
      const r = await enrichType({
        typeKey: t,
        dump: pendingDir(opts.dump),
        http: opts.http,
        githubToken: opts.githubToken,
        concurrency: opts.concurrency,
        delayMs: opts.delayMs,
        limit: opts.limit,
        refetch: false,
        refetchErrors: false,
        logger: opts.logger,
        sleep: opts.sleep,
      });
      if (!r.missingDir && (r.enriched || r.failed)) {
        logger.info(`  [${t}] (_pending) filled ${r.enriched} body(ies), ${r.failed} failed`);
      }
    }
  }

  // Record any --refetch overwrites that were staged this run.
  if (staged.length) await mergePending(opts.dump, staged, new Date().toISOString());

  // Write a machine-readable report so a long driven run can be inspected and
  // re-run (re-process the failures with --refetch-errors).
  const reportPath = `${opts.dump}/_enrich_report.json`;
  const totalFailed = reports.reduce((a, r) => a + r.failed, 0);
  try {
    await writeJson(reportPath, { generatedAt: new Date().toISOString(), types: reports });
    logger.info(`\nReport: ${reportPath}`);
  } catch (e) {
    logger.warn(`Could not write report: ${(e as Error).message}`);
  }
  if (totalFailed) {
    logger.warn(
      `\nFAILURES (${totalFailed}): ` +
        reports.filter((r) => r.failed).map((r) => `${r.typeKey}=${r.failed}`).join(", ") +
        ` — see ${reportPath}; re-run with --refetch-errors after fixing.`,
    );
  }
  logger.info("All done.");
  return reports;
}

File: f5kb/enrich/driver.py
"""Enrichment driver: walks a dump, runs per-type enrichers with a thread pool."""

from __future__ import annotations

import datetime
import time
from concurrent.futures import ThreadPoolExecutor, as_completed
from dataclasses import dataclass, field
from pathlib import Path
from threading import Lock

from f5kb.enrich.enrichers import STALE_KEYS, TYPE_ENRICHERS, has_body
from f5kb.http.fetcher import HttpClient
from f5kb.lib.fsutil import path_exists, read_json, walk_article_files, write_json
from f5kb.lib.logger import NULL_LOGGER, Logger
from f5kb.lib.staging import PendingEntry, merge_pending, pending_dir, pending_path


@dataclass
class TypeReport:
    type_key: str
    files: int = 0
    enriched: int = 0
    failed: int = 0
    skipped: int = 0
    staged: int = 0
    missing_dir: bool = False
    errors: list[dict] = field(default_factory=list)


def enrich_type(
    type_key: str,
    dump: str,
    http: HttpClient,
    *,
    github_token: str | None = None,
    concurrency: int = 4,
    delay_ms: int = 0,
    limit: int | None = None,
    refetch: bool = False,
    refetch_errors: bool = False,
    logger: Logger = NULL_LOGGER,
    sleep=None,
    changelog=None,
    approval: bool = False,
    pending: list[PendingEntry] | None = None,
) -> TypeReport:
    if sleep is None:
        sleep = time.sleep
    log = logger
    report = TypeReport(type_key=type_key)
    enricher = TYPE_ENRICHERS.get(type_key)
    if not enricher:
        log.info(f"  [{type_key}] no enricher implemented — skipping")
        return report

    type_dir = str(Path(dump) / type_key)
    try:
        files = sorted(walk_article_files(type_dir))
    except Exception:
        log.info(f"  [{type_key}] no directory {type_dir} — skipping")
        report.missing_dir = True
        return report

    if limit is not None:
        files = files[:limit]
    report.files = len(files)

    now_iso = datetime.datetime.now(datetime.timezone.utc).isoformat().replace("+00:00", "Z")
    done_count = 0
    lock = Lock()

    def _worker(file_path: str):
        nonlocal done_count
        article = read_json(file_path)
        had_error = isinstance((article.get("content") or {}).get("bodyError"), str)
        if not refetch and not (refetch_errors and had_error) and has_body(article.get("content")):
            with lock:
                report.skipped += 1
            return

        had_body_before = has_body(article.get("content"))
        try:
            result = enricher(article, now_iso, http, github_token=github_token)
            with lock:
                report.enriched += 1
        except Exception as e:
            error = str(e)
            result = {"bodySource": article.get("link") or "", "fetchedAt": now_iso, "bodyError": error}
            with lock:
                report.failed += 1
                report.errors.append({"id": article.get("id") or "", "link": article.get("link") or "", "error": error})

        if changelog:
            op = "body-error" if result.get("bodyError") else ("body-changed" if had_body_before else "body-added")
            changelog.record(op, article.get("documentType") or type_key, article.get("id") or "",
                             title=article.get("title"), source="enrich", detail=result.get("bodyError"))

        base = dict(article.get("content") or {})
        for k in STALE_KEYS:
            base.pop(k, None)
        article["content"] = {**base, **result}

        art_id = article.get("id") or ""
        if approval and refetch and had_body_before:
            pp = pending_path(dump, type_key, art_id)
            Path(pp).parent.mkdir(parents=True, exist_ok=True)
            write_json(pp, article)
            if pending is not None:
                with lock:
                    pending.append(PendingEntry(
                        type_key=type_key, id=art_id, title=article.get("title"),
                        op="edited", source="enrich", staged_at=now_iso,
                    ))
            with lock:
                report.staged += 1
        else:
            write_json(file_path, article)

        with lock:
            done_count += 1
            total_done = done_count + report.skipped
            if total_done % 25 == 0:
                log.info(f"  [{type_key}] {total_done}/{report.files} "
                         f"(ok={report.enriched} fail={report.failed} skip={report.skipped})")
            if delay_ms:
                sleep(delay_ms / 1000)

    effective_concurrency = min(concurrency, max(1, len(files)))
    with ThreadPoolExecutor(max_workers=effective_concurrency) as ex:
        futures = [ex.submit(_worker, f) for f in files]
        for fut in as_completed(futures):
            try:
                fut.result()
            except Exception as e:
                log.warn(f"  [{type_key}] worker error: {e}")

    stg = f" staged={report.staged}" if report.staged else ""
    log.info(f"  [{type_key}] DONE: {report.files} files — "
             f"enriched={report.enriched} failed={report.failed} skipped={report.skipped}{stg}")
    return report


def enrich_dump(
    dump: str,
    *,
    types: list[str] | None = None,
    exclude_types: list[str] | None = None,
    http: HttpClient,
    github_token: str | None = None,
    concurrency: int = 4,
    delay_ms: int = 0,
    limit: int | None = None,
    refetch: bool = False,
    refetch_errors: bool = False,
    logger: Logger = NULL_LOGGER,
    sleep=None,
    changelog=None,
    approval: bool = False,
) -> list[TypeReport]:
    log = logger
    requested = types or list(TYPE_ENRICHERS.keys())
    excluded = set(exclude_types or [])
    to_run = [t for t in requested if t not in excluded and TYPE_ENRICHERS.get(t)]
    for t in requested:
        if t not in excluded and not TYPE_ENRICHERS.get(t):
            log.error(f"  [{t}] no enricher implemented — skipping")

    if not to_run:
        raise ValueError("Nothing to do. Implemented types: " + ", ".join(TYPE_ENRICHERS.keys()))

    log.info(f"Enriching bodies in {dump} for: {', '.join(to_run)}")
    log.info(f"(concurrency={concurrency}, delay={delay_ms}ms, refetch={refetch})")
    if "F5_GitHub" in to_run:
        _gh_auth = "token present (5000/hr)" if github_token else "UNAUTHENTICATED (60/hr) — set GITHUB_TOKEN to raise"
        log.info(f"GitHub auth: {_gh_auth}")

    staged: list[PendingEntry] = []
    reports: list[TypeReport] = []
    for t in to_run:
        reports.append(enrich_type(
            t, dump, http,
            github_token=github_token,
            concurrency=concurrency,
            delay_ms=delay_ms,
            limit=limit,
            refetch=refetch,
            refetch_errors=refetch_errors,
            logger=logger,
            sleep=sleep,
            changelog=changelog,
            approval=approval,
            pending=staged,
        ))

    # Also enrich articles in _pending/ (so reviewers see complete new articles)
    pdir = pending_dir(dump)
    if path_exists(pdir):
        for t in to_run:
            r = enrich_type(
                t, pdir, http,
                github_token=github_token,
                concurrency=concurrency,
                delay_ms=delay_ms,
                limit=limit,
                refetch=False,
                refetch_errors=False,
                logger=logger,
                sleep=sleep,
            )
            if not r.missing_dir and (r.enriched or r.failed):
                log.info(f"  [{t}] (_pending) filled {r.enriched} body(ies), {r.failed} failed")

    if staged:
        now_iso = datetime.datetime.now(datetime.timezone.utc).isoformat().replace("+00:00", "Z")
        merge_pending(dump, staged, now_iso)

    report_path = str(Path(dump) / "_enrich_report.json")
    total_failed = sum(r.failed for r in reports)
    try:
        write_json(report_path, {
            "generatedAt": datetime.datetime.now(datetime.timezone.utc).isoformat().replace("+00:00", "Z"),
            "types": [
                {
                    "typeKey": r.type_key,
                    "files": r.files,
                    "enriched": r.enriched,
                    "failed": r.failed,
                    "skipped": r.skipped,
                    "staged": r.staged,
                    "errors": r.errors,
                }
                for r in reports
            ],
        })
        log.info(f"\nReport: {report_path}")
    except Exception as e:
        log.warn(f"Could not write report: {e}")

    if total_failed:
        failed_parts = ", ".join(f"{r.type_key}={r.failed}" for r in reports if r.failed)
        log.warn(f"\nFAILURES ({total_failed}): {failed_parts} — see {report_path}; "
                 "re-run with --refetch-errors after fixing.")

    log.info("All done.")
    return reports
